### PR TITLE
MRG: Add minimum_phase filter function

### PR DESCRIPTION
**Pull request by larsoner**
Adds a function for converting a linear-phase FIR filters to minimum phase. Currently uses ~~one~~ two of four possible methods (Hilbert _and homomorphic_)~~, and probably not the most general one, at least not based on my implementation. I plan to also implement the homomorphic method.~~

Includes tests and an example, so people can feel free to look and comment if they have time.

### CODE REVIEW

**Comment by Eric89GXL** [comment]
Ready for review/merge from my end. CIs are happy and rendered docs are here:

https://800-1460385-gh.circle-artifacts.com/0//home/ubuntu/scipy/doc/build/html-scipyorg/generated/scipy.signal.minimum_phase.html#scipy.signal.minimum_phase

@WarrenWeckesser it looks like you did a lot of the FIR filter design stuff (over 5 years ago!) so you might be interested in this.

**Comment by WarrenWeckesser** [comment]
I'm interested, but I might not get to a review for several days (and in Warren time, that can be a lot longer).  So if any of the other active `signal` folks get involved, don't wait for me.

**Comment by Eric89GXL** [comment]
There is no big rush, would be good to get for 0.19 though.

**Comment by Eric89GXL** [comment]
I take it Warren time has taken over @WarrenWeckesser... :)

@e-q @endolith any interest in reviewing filter design stuff like this?

File: scipy/signal/fir_filter_design.py [hunk 1]
```diff
@@ -791,3 +799,201 @@ def firls(numtaps, bands, desired, weight=None, nyq=1.):
...
+        preferably fast for FFT/IFFT.
+    """
+    # Adapted based on code by Niranjan Damera-Venkata,
+    # Brian L. Evans and Shawn R. McCaslin
+    sig = np.zeros(len(mag))
+    # Leave Nyquist and DC at 0, knowing np.abs(fftfreq(N)[midpt]) == 0.5
+    midpt = len(mag) // 2
+    sig[1:midpt] = 1
+    sig[midpt+1:] = -1
+    recon = ifft(mag * np.exp(fft(sig * ifft(np.log(np.abs(mag))))))
```
Comment (endolith) [inline scipy/signal/fir_filter_design.py]: why `abs(mag)` if mag is already a magnitude spectrum?

File: scipy/signal/fir_filter_design.py [hunk 2]
```diff
@@ -791,3 +799,201 @@ def firls(numtaps, bands, desired, weight=None, nyq=1.):
...
+    See Also
+    --------
+    firwin
+    firwin2
+    remez
+
+    Notes
+    -----
+    Both the Hilbert [1]_ or homomorphic [4]_ [5]_ methods require
+    selection of an FFT length to estimate the for the complex cepstrum
```
Comment (endolith) [inline scipy/signal/fir_filter_design.py]: typo "the for the"

File: scipy/signal/fir_filter_design.py [hunk 3]
```diff
@@ -791,3 +799,201 @@ def firls(numtaps, bands, desired, weight=None, nyq=1.):
...
+           Upper Saddle River, N.J.: Prentice Hall, 1999.
+    """  # noqa
+    h = np.asarray(h)
+    if np.iscomplexobj(h):
+        raise ValueError('Complex filters not supported yet')
+    if h.ndim != 1 or h.size <= 2:
+        raise ValueError('h must be 1D and at least 2 samples long')
+    n_half = len(h) // 2
+    if not np.allclose(h[-n_half:][::-1], h[:n_half]):
+        warnings.warn('h does not appear to by symmetric, conversion may '
```
Comment (endolith) [inline scipy/signal/fir_filter_design.py]: this needs a test case

File: scipy/signal/fir_filter_design.py [hunk 4]
```diff
@@ -791,3 +799,201 @@ def firls(numtaps, bands, desired, weight=None, nyq=1.):
...
+                      'fail')
+    if not isinstance(method, string_types) or method not in \
+            ('homomorphic', 'hilbert',):
+        raise ValueError('method must be "homomorphic" or "hilbert", got %r'
```
Comment (endolith) [inline scipy/signal/fir_filter_design.py]: this needs a test case (invalid method)

**Comment by Eric89GXL** [comment]
Comments addressed, thanks @endolith

File: scipy/signal/fir_filter_design.py [hunk 5]
```diff
@@ -791,3 +799,200 @@ def firls(numtaps, bands, desired, weight=None, nyq=1.):
...
     return coeffs
+
+
+def _dhtm(mag):
+    """Compute the modified 1D discrete Hilbert transform
+
+    Parameters
+    ----------
+    mag : ndarray
+        The magnitude spectrum. Should be an even number, and
```
Comment (e-q) [inline scipy/signal/fir_filter_design.py]: Do you mean that the length of the array should be an even number?

File: scipy/signal/fir_filter_design.py [hunk 6]
```diff
@@ -791,3 +799,200 @@ def firls(numtaps, bands, desired, weight=None, nyq=1.):
...
+        preferably fast for FFT/IFFT.
+    """
+    # Adapted based on code by Niranjan Damera-Venkata,
+    # Brian L. Evans and Shawn R. McCaslin
```
Comment (e-q) [inline scipy/signal/fir_filter_design.py]: Is there a package name / paper / link you could add for this reference and add that reference to the docstring?

File: scipy/signal/fir_filter_design.py [hunk 7]
```diff
@@ -791,3 +799,200 @@ def firls(numtaps, bands, desired, weight=None, nyq=1.):
...
+           New York: Wiley-Interscience, 1993.
+    .. [4] J. S. Lim, Advanced Topics in Signal Processing.
+           Englewood Cliffs, N.J.: Prentice Hall, 1988.
+    .. [5] A. V. Oppenheim, R. W. Schafer, and J. R. Buck,
+           "Discrete-Time Signal Processing," 2nd edition.
+           Upper Saddle River, N.J.: Prentice Hall, 1999.
+    """  # noqa
+    h = np.asarray(h)
+    if np.iscomplexobj(h):
+        raise ValueError('Complex filters not supported yet')
```
Comment (e-q) [inline scipy/signal/fir_filter_design.py]: I'd omit the "yet". Under promise and over deliver :)

**Review by e-q — APPROVED** [review]
Made some minor nitpicky comments, but looks good to me!

**Comment by Eric89GXL** [comment]
Thanks @e-q comments addressed.

**Comment by e-q** [comment]
Everyone seems happy, I'll pull the trigger in a few days if nothing else crops up.

**Pull request merged**